File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/HeartRateVariabilityRmssdRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.HeartRateVariabilityRmssdRecord
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.HeartRateVariabilityRMSSDRecordDto
import java.time.Instant
import java.time.ZoneOffset

internal fun HeartRateVariabilityRmssdRecord.toDto(): HeartRateVariabilityRMSSDRecordDto =
    HeartRateVariabilityRMSSDRecordDto(
        id = metadata.id,
        time = time.toEpochMilli(),
        zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
        metadata = metadata.toDto(),
        heartRateVariabilityMillis = heartRateVariabilityMillis,
    )

internal fun HeartRateVariabilityRMSSDRecordDto.toHealthConnect(): HeartRateVariabilityRmssdRecord =
    HeartRateVariabilityRmssdRecord(
        heartRateVariabilityMillis = heartRateVariabilityMillis,
        time = Instant.ofEpochMilli(time),
        zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/heart_rate/heart_rate_variability_rmssd_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show HeartRateVariabilityRMSSDRecord, HealthRecordId, Number, sinceV2_2_0;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show HeartRateVariabilityRMSSDRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [HeartRateVariabilityRMSSDRecord] to
/// [HeartRateVariabilityRMSSDRecordDto].
@sinceV2_2_0
@internal
extension HeartRateVariabilityRMSSDRecordToDto
    on HeartRateVariabilityRMSSDRecord {
  HeartRateVariabilityRMSSDRecordDto toDto() {
    return HeartRateVariabilityRMSSDRecordDto(
      id: id.toDto(),
      time: time.millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      heartRateVariabilityMillis: heartRateVariabilityMillis.value.toDouble(),
    );
  }
}

/// Converts [HeartRateVariabilityRMSSDRecordDto] to
/// [HeartRateVariabilityRMSSDRecord].
@sinceV2_2_0
@internal
extension HeartRateVariabilityRMSSDRecordDtoToDomain
    on HeartRateVariabilityRMSSDRecordDto {
  HeartRateVariabilityRMSSDRecord toDomain() {
    return HeartRateVariabilityRMSSDRecord(
      id: id?.toDomain() ?? HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      heartRateVariabilityMillis: Number(heartRateVariabilityMillis),
    );
  }
}
